lib/taste_tester/commands.rb

```diff
@@ -102,29 +102,47 @@ def self.test
         TasteTester::Hooks.pre_test(TasteTester::Config.dryrun, repo, hosts)
       end
       tested_hosts = []
+      host_threads = []
+      connect_failures = 0
       hosts.each do |hostname|
-        host = TasteTester::Host.new(hostname, server)
+        # Poor man thread pool manager: keeping it simple
+        nb_threads_over_max = host_threads.length - TasteTester::Config.parallel_hosts
+        if nb_threads_over_max >= 0
+          host_threads[nb_threads_over_max].join
+        end
```

Comment:
Wait, so if we have 10 max threads, and we've made an 11th, we wait for the 11th to finish? That's... I mean that'll solve for too-many-threads for sure. But it also means that if you have 20 hosts to test and a max conns of 10, you'll do 10 in parallel, then you'll wait for the 11th (which means the other 10 are likely to finish up), and then proceed to do the next 10 one at a time (since the number of elements in the array is still over 10). You'll fire off 12, then wait for 12, then fire off 13 then wait for 13 and so on. That doesn't seem like it does what you want.

Reply:
No. It's not pure thread pool, 'cause I didn't want to use a Queue (I could, that would make @rb2k happy).
Basically, if you have 10 threads and you're trying to put the 11th, you'll wait for the **first** one to finish. Then when it finishes, the 12th will wait for the second one, etc...
Yes, in practice, it could very well be that you're waiting on the one thread that's taking the longest, and you could run more. The added complexity did not feel worth it, but I don't mind creating a Queue and having N threads polling from it instead.

Reply:
OK, I caved and implemented a thread pool like @rb2k suggested. Still works pretty much the same in my testing :)